core/common/src/Clock.kt

```diff
@@ -53,5 +53,19 @@ private class InstantTimeMark(private val instant: Instant, private val clock: C
     override fun minus(duration: Duration): TimeMark = InstantTimeMark(instant - duration, clock)
 }
 
+/**
+ * Returns the [Clock] by storing an initial [TimeMark] using [TimeSource.markNow] and [returns][Clock.now] the elapsed
+ * time using [TimeMark.elapsedNow] plus the provided [offset].
+ *
+ * This clock stores the initial [TimeMark], so repeatedly creating [Clock]s from the same [TimeSource] results
+ * into different [Instant]s iff the time of the [TimeSource] was increased. To sync different [Clock]s, use the [offset]
+ * parameter.
+ */
+@ExperimentalTime
+public fun TimeSource.asClock(offset: Instant = Instant.fromEpochSeconds(0)): Clock = object : Clock {
```

Comment:
We've deliberated internally about this function, and it does seem useful. A couple of notes:
* The name `offset` leaves the impression that it's somehow related to `UtcOffset`, but it's not. We arrived at the name `origin` instead.
* Having a default parameter here leaves the wrong impression. `TimeSource.Monotonic.asClock()` seems benign, but starts returning dates far in the past. `Clock.System.now()` would be less surprising, but it's nondeterministic, which is a bad fit for tests (which we'd expect to be the most common use case).